adapters/waardex/waardex.go

```diff
@@ -0,0 +1,288 @@
+package waardex
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"text/template"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v3/adapters"
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/macros"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+)
+
+type waardexAdapter struct {
+	EndpointTemplate *template.Template
+}
+
+// MakeRequests prepares request information for prebid-server core
+func (adapter *waardexAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		errs = append(errs, newBadInputError("No impression in the bid request"))
+		return nil, errs
+	}
+	imps, impExts, err := getImpressionsInfo(request.Imp)
+	if len(imps) == 0 {
+		return nil, err
+	}
+	errs = append(errs, err...)
+
+	pub2impressions, dispErrors := dispatchImpressions(imps, impExts)
```

Comment:
Nitpick: We can combine these two functions `getImpressionsInfo` and  `dispatchImpressions`. This will avoid multiple for loops and will have single impression related utility which can be easy to maintain